File: themes/n-queens/python/src/n_queens/solver.py
"""Backtracking solver for the N-Queens problem."""

from __future__ import annotations


class NQueensError(Exception):
    """Raised when n-queens cannot complete successfully."""


def is_safe(columns: list[int], row: int, col: int) -> bool:
    """Check whether placing a queen at (row, col) conflicts with rows above it."""
    for placed_row in range(row):
        placed_col = columns[placed_row]
        if placed_col == col or abs(placed_col - col) == abs(placed_row - row):
            return False
    return True


def solve(n: int) -> list[list[int]]:
    """Return all N-Queens solutions as column-index arrays, in discovery order."""
    if n < 1:
        raise NQueensError(f"n must be >= 1, got {n}")

    results: list[list[int]] = []
    columns = [0] * n

    def backtrack(row: int) -> None:
        if row == n:
            results.append(columns.copy())
            return
        for col in range(n):
            if is_safe(columns, row, col):
                columns[row] = col
                backtrack(row + 1)

    backtrack(0)
    return results


def format_solutions(solutions: list[list[int]]) -> str:
    """Format solutions as one line per solution plus a total-count line."""
    lines = [" ".join(map(str, solution)) for solution in solutions]
    lines.append(f"Total solutions: {len(solutions)}")
    return "\n".join(lines) + "\n"

File: themes/n-queens/typescript/src/solver.ts
/**
 * Backtracking solver for the N-Queens problem.
 */

export class NQueensError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "NQueensError";
  }
}

export function isSafe(columns: number[], row: number, col: number): boolean {
  for (let placedRow = 0; placedRow < row; placedRow++) {
    const placedCol = columns[placedRow];
    if (placedCol === col || Math.abs(placedCol - col) === Math.abs(placedRow - row)) {
      return false;
    }
  }
  return true;
}

export function solve(n: number): number[][] {
  if (n < 1) {
    throw new NQueensError(`n must be >= 1, got ${n}`);
  }

  const results: number[][] = [];
  const columns: number[] = new Array(n).fill(0);

  function backtrack(row: number): void {
    if (row === n) {
      results.push([...columns]);
      return;
    }
    for (let col = 0; col < n; col++) {
      if (isSafe(columns, row, col)) {
        columns[row] = col;
        backtrack(row + 1);
      }
    }
  }

  backtrack(0);
  return results;
}

export function formatSolutions(solutions: number[][]): string {
  const lines = solutions.map((solution) => solution.join(" "));
  lines.push(`Total solutions: ${solutions.length}`);
  return `${lines.join("\n")}\n`;
}
